plugin/federation/federation.go

```diff
@@ -0,0 +1,335 @@
+	cfg.Directives["requires"] = config.DirectiveConfig{SkipRuntime: true}
+	cfg.Directives["provides"] = config.DirectiveConfig{SkipRuntime: true}
+	cfg.Directives["key"] = config.DirectiveConfig{SkipRuntime: true}
+	cfg.Directives["extends"] = config.DirectiveConfig{SkipRuntime: true}
+
+	return nil
+}
+
+func (f *federation) InjectSources(cfg *config.Config) {
+	cfg.AdditionalSources = append(cfg.AdditionalSources, f.getSource(false))
+	f.setEntities(cfg)
+	s := "type Entity {\n"
+	for _, e := range f.Entities {
+		s += fmt.Sprintf("\t%s(%s: %s): %s!\n", e.ResolverName, e.FieldName, e.FieldTypeGQL, e.Name)
+	}
+	s += "}"
+	cfg.AdditionalSources = append(cfg.AdditionalSources, &ast.Source{Name: "entity.graphql", Input: s, BuiltIn: true})
+}
+
+func (f *federation) MutateSchema(s *ast.Schema) error {
+	// --- Set _Entity Union ---
+	union := &ast.Definition{
+		Name:        "_Entity",
+		Kind:        ast.Union,
+		Description: "A union unifies all @entity types (TODO: interfaces)",
+		Types:       []string{},
+	}
+	for _, ent := range f.Entities {
+		union.Types = append(union.Types, ent.Name)
+		s.AddPossibleType("_Entity", ent.Def)
+		// s.AddImplements(ent.Name, union) // Do we need this?
+	}
+	s.Types[union.Name] = union
+
+	// --- Set _entities query ---
+	fieldDef := &ast.FieldDefinition{
+		Name: "_entities",
+		Type: ast.NonNullListType(ast.NamedType("_Entity", nil), nil),
+		Arguments: ast.ArgumentDefinitionList{
+			{
+				Name: "representations",
+				Type: ast.NonNullListType(ast.NonNullNamedType("_Any", nil), nil),
+			},
+		},
+	}
+	if s.Query == nil {
+		s.Query = &ast.Definition{
+			Kind: ast.Object,
+			Name: "Query",
+		}
+		s.Types["Query"] = s.Query
+	}
+	s.Query.Fields = append(s.Query.Fields, fieldDef)
+
+	// --- set _Service type ---
+	typeDef := &ast.Definition{
+		Kind: ast.Object,
+		Name: "_Service",
+		Fields: ast.FieldList{
+			&ast.FieldDefinition{
+				Name: "sdl",
+				Type: ast.NonNullNamedType("String", nil),
+			},
+		},
+	}
+	s.Types[typeDef.Name] = typeDef
+
+	// --- set _service query ---
+	_serviceDef := &ast.FieldDefinition{
+		Name: "_service",
+		Type: ast.NonNullNamedType("_Service", nil),
+	}
+	s.Query.Fields = append(s.Query.Fields, _serviceDef)
+	return nil
+}
+
+func (f *federation) getSource(builtin bool) *ast.Source {
+	return &ast.Source{
+		Name: "federation.graphql",
+		Input: `# Declarations as required by the federation spec 
+# See: https://www.apollographql.com/docs/apollo-server/federation/federation-spec/
+
+scalar _Any
+scalar _FieldSet
+
+directive @external on FIELD_DEFINITION
+directive @requires(fields: _FieldSet!) on FIELD_DEFINITION
+directive @provides(fields: _FieldSet!) on FIELD_DEFINITION
+directive @key(fields: _FieldSet!) on OBJECT | INTERFACE
+directive @extends on OBJECT
+`,
+		BuiltIn: builtin,
+	}
+}
+
+// Entity represents a federated type
+// that was declared in the GQL schema.
+type Entity struct {
```

Comment:
I think I would prefer if this referenced the underlying types where possible, rather than copying them just out, just in case another plugin is modifying them.
 - We already have `Def`, so `Name` is redundant
 - FieldName and FieldTypeGQL come from `FieldDef`, so let's add that instead?

If we apply the same logic to `RequireField` I think it can be replaced entirely by a `codegen.Field`

Reply:
@vektah I did it for `Def` and `FieldDef`. However, `RequireField` would make `setEntities` create a dummy `codegen.Field` so I can have a Name to match against in `GenerateCode`, then `GenerateCode` will have to replace its codegen.Field with the dummy one which fields a bit ugly. Would you like me to switch that up as well?